linkerd/proxy/pool/src/worker.rs

```diff
@@ -0,0 +1,349 @@
+                biased;
+
+                // If the pool updated, continue waiting for the pool to be
+                // ready.
+                res = self.discovery.discover() => {
+                    let update = res?;
+                    tracing::debug!(?update, "Discovered");
+                    self.pool.pool.update_pool(update);
+                }
+
+                // When the pool is ready, clear any failfast state we may have
+                // set before returning.
+                res = self.pool.ready() => {
+                    tracing::trace!(ready.ok = res.is_ok());
+                    return res;
+                }
+            }
+        }
+    }
+}
+
+// === impl Discovery ===
+
+impl<T, R> Discovery<R>
+where
+    T: Clone + Eq + std::fmt::Debug,
+    R: TryStream<Ok = Update<T>> + Unpin,
+    R::Error: Into<Error>,
+{
+    fn new(resolution: R) -> Self {
+        Self {
+            resolution,
+            closed: false,
+        }
+    }
+
+    /// Await the next service discovery update.
+    ///
+    /// If the discovery stream has closed, this never returns.
+    async fn discover(&mut self) -> Result<Update<T>, Error> {
+        if self.closed {
+            // Never returns.
+            return futures::future::pending().await;
+        }
+
+        match self.resolution.try_next().await {
+            Ok(Some(up)) => Ok(up),
+
+            Ok(None) => {
+                tracing::debug!("Resolution stream closed");
+                self.closed = true;
+                // Never returns.
+                futures::future::pending().await
+            }
+
+            Err(e) => {
+                let error = e.into();
+                tracing::debug!(%error, "Resolution stream failed");
+                self.closed = true;
+                Err(error)
+            }
+        }
+    }
+}
+
+// === impl PoolDriver ===
+
+impl<P> PoolDriver<P> {
+    fn new(pool: P, failfast: Failfast) -> Self {
+        Self { pool, failfast }
+    }
+
+    /// Drives all current endpoints to ready.
+    ///
+    /// If the service is in failfast, this clears the failfast state on readiness.
+    ///
+    /// If any endpoint fails, the error
+    /// is returned.
+    async fn drive<T, Req>(&mut self) -> Error
+    where
+        P: Pool<T, Req>,
+        P::Error: Into<Error>,
+    {
+        if self.failfast.is_active() {
+            tracing::trace!("Waiting to leave failfast");
+            let res = self.pool.ready().await;
+            match self.failfast.set_ready() {
+                Some(failfast::State::Failfast { since }) => {
+                    tracing::info!(
+                        elapsed = (time::Instant::now() - since).as_secs_f64(),
+                        "Available; exited failfast"
+                    );
+                }
+                _ => unreachable!("must be in failfast"),
+            }
+            if let Err(e) = res {
+                return e.into();
+            }
```

Comment:
tiny style nit: isn't this just
```suggestion
            res?;
```